scipy/stats/_binned_statistic.py

```diff
@@ -541,55 +544,20 @@ def binned_statistic_dd(sample, values, statistic='mean',
     except TypeError:
         bins = Ndim * [bins]
 
-    # Select range for each dimension
-    # Used only if number of bins is given.
-    if range is None:
-        smin = np.atleast_1d(np.array(sample.min(axis=0), float))
-        smax = np.atleast_1d(np.array(sample.max(axis=0), float))
-    else:
-        smin = np.zeros(Ndim)
-        smax = np.zeros(Ndim)
-        for i in xrange(Ndim):
-            smin[i], smax[i] = range[i]
-
-    # Make sure the bins have a finite width.
-    for i in xrange(len(smin)):
-        if smin[i] == smax[i]:
-            smin[i] = smin[i] - .5
-            smax[i] = smax[i] + .5
-
     # Create edge arrays
-    for i in xrange(Ndim):
-        if np.isscalar(bins[i]):
-            nbin[i] = bins[i] + 2  # +2 for outlier bins
-            edges[i] = np.linspace(smin[i], smax[i], nbin[i] - 1)
-        else:
-            edges[i] = np.asarray(bins[i], float)
-            nbin[i] = len(edges[i]) + 1  # +1 for outlier bins
-        dedges[i] = np.diff(edges[i])
-
-    nbin = np.asarray(nbin)
+    if reuse is None:
+        nbin, edges, dedges = _bin_edges(sample, bins, range)
+    else:
+        edges = reuse.bin_edges
+        nbin = np.array([len(edges[i]) + 1 for i in xrange(Ndim)])
+        # +1 for outlier bins
+        dedges = [np.diff(edges[i]) for i in xrange(Ndim)]
 
     # Compute the bin number each sample falls into, in each dimension
-    sampBin = [
-        np.digitize(sample[:, i], edges[i])
-        for i in xrange(Ndim)
-    ]
-
-    # Using `digitize`, values that fall on an edge are put in the right bin.
-    # For the rightmost bin, we want values equal to the right
-    # edge to be counted in the last bin, and not as an outlier.
-    for i in xrange(Ndim):
-        # Find the rounding precision
-        decimal = int(-np.log10(dedges[i].min())) + 6
-        # Find which points are on the rightmost edge.
-        on_edge = np.where(np.around(sample[:, i], decimal) ==
-                           np.around(edges[i][-1], decimal))[0]
-        # Shift these points one bin to the left.
-        sampBin[i][on_edge] -= 1
-
-    # Compute the sample indices in the flattened statistic matrix.
-    binnumbers = np.ravel_multi_index(sampBin, nbin)
+    if reuse is None:
```

Comment:
the if/else block above has the same logic e.g. `if resue is None:` etc. why can't all the logic be in 1 if/else block rather than 2?

Reply:
The two blocks have been merged